Headless / API usage docs improvements (#13084)
wagtail/wagtail: docs/advanced_topics/api/django-ninja.md

        model = BlogPage
        model_fields = [
            "intro",
        ]

@api.get("/pages/{page_id}/", response=BlogPageSchema)
def get_page(request: "HttpRequest", page_id: int):
    return get_object_or_404(Page, id=page_id).specific
```

This works well, with the endpoint now returning generic `Page` fields and the `BlogPage` introduction.
But for sites where all page content is served via an API, it could become tedious to create new endpoints for every page type.

### Combining multiple schemas

To reflect our response may return multiple page types, we use the `Union` type hint to combine multiple schemas.
This allows us to return different page types from the same endpoint.
Here is an example with an additional schema for our `HomePage` type:

```python
class HomePageSchema(BasePageSchema, ModelSchema):
    class Config(BasePageSchema.Config):
        model = HomePage

@api.get("/pages/{page_id}/", response=Union[BlogPageSchema, HomePageSchema])
def get_page(request: "HttpRequest", page_id: int):
    return get_object_or_404(Page, id=page_id).specific
```

With this in place, we are still missing a way to determine which of the schemas to use for a given page.

> We want to do this by page type, adding an extra `content_type` parameter to our schemas.

Author: Ah good catch.

- For `BasePageSchema`, we define `content_type: str`, as any page type can use this base.
- For `HomePageSchema`, we set `content_type: Literal["homepage"]`.

> - And for `BlogPageSchema`, we set `content_type: Literal["blogpage"]`.

Author: Let’s!

All we need now is to add a [resolver](https://django-ninja.dev/guides/response/#resolvers) calculated field to the `BasePageSchema`, to return the correct content type for each page.

```python
    @staticmethod
    def resolve_content_type(page: Page) -> str:

>         return page.content_type.model

Author: Ty!

```

With this in place, Pydantic is able to validate the page data returned in `get_page` according to one of the schemas in the `response` Union.
It then serializes the data according to the specific schema.

### Nested data

Where the page schema references data in separate models, rather than creating new endpoints, we can add the data directly to the page schema.
Here is an example, adding blog page authors (a snippet with a `ParentalManyToManyField`):

```python
class BlogPageSchema(BasePageSchema, ModelSchema):
    content_type: Literal["blogpage"]
    authors: List[str] = []

    class Config(BasePageSchema.Config):
        model = BlogPage
        model_fields = [
            "intro",
        ]

    @staticmethod
    def resolve_authors(page: BlogPage, context) -> List[str]:
        return [author.name for author in page.authors.all()]
```

> This could also be done with the Pydantic `Field` function if the `BlogPage` class had a method to retrieve author names directly: `authors: List[str] = Field([], alias="get_author_names")`.

Author: Thank you!

### Rich text in the API

Rich text fields in Wagtail use a specific internal format, described in [](../../extending/rich_text_internals). They can be added to the schema as `str`, but it’s often more useful for the API to provide a “display” representation, where references to pages and images are replaced with URLs.

> This can also be done with [Ninja resolvers](https://django-ninja.dev/guides/response/#resolvers). Here is an example with the `HomePageSchema`:

Author: I thought it would be worthwhile to link again to make this more self-contained

```python
from wagtail.images.models import AbstractRendition

class RenditionSchema(ModelSchema):
    # We need to use the Field / alias API for properties
    url: str = Field(None, alias="file.url")
    alt: str = Field(None, alias="alt")

    class Config:
        model = AbstractRendition
        model_fields = [
            "width",
            "height",
        ]
```

On the `BlogPageSchema`, we define our image field as: `main_image: List[RenditionSchema] = []`. Then add the resolver for it:

```python
    @staticmethod
    def resolve_main_image(page: BlogPage) -> List[AbstractRendition]:
        filters = [
            "fill-800x600|format-webp",
            "fill-800x600",
        ]
        if image := page.main_image():
            return image.get_renditions(*filters).values()
        return []
```

In JSON, our `main_image` is now represented as an array, where each item is an object with `url`, `alt`, `width`, and `height` properties.

## OpenAPI documentation

Django Ninja generates [OpenAPI](https://swagger.io/specification/) documentation automatically, based on the defined operations and schemas.

> This also includes a documentation viewer, with support to try out the API directly from the browser.

Reviewer: Maybe add a suggestion to check it out?

```suggestion
This also includes a documentation viewer, with support to try out the API directly from the browser. With the above example, you can try accessing the docs at `/api/docs`.
```

(Annoyingly, the path doesn't have a trailing slash and adding it results in 404 in my case for some reason.)

Author: Let’s. This is already an early step in the mini tutorial at the top of the page but certainly worth mentioning again.

(api_ninja)=

# How to set up Django Ninja

> Even though Wagtail provides a [built-in API module](api) based on Django REST Framework, it is possible to use other API frameworks.

Reviewer 2: ```suggestion
While Wagtail provides a [built-in API module](api) based on Django REST Framework, it is possible to use other API frameworks.
```

'Even though' makes it sound like it should be a surprise :)

```

Ninja treats every parameter of the `list_pages` function as a query parameter. It uses the provided type hint to parse the value, validate it, and generate the OpenAPI schema.

### Adding custom page fields

Next, let’s add a "detail" API endpoint to return a single page of a specific type. We can use the [path parameters](https://django-ninja.dev/guides/input/path-params/) from Ninja to retrieve our `page_id`.
We also create a new schema for a specific page type: here, `BlogPage`, with `BasePageSchema` as a base.

```python
from blog.models import BlogPage

class BlogPageSchema(BasePageSchema, ModelSchema):
    class Config(BasePageSchema.Config):
        model = BlogPage
        model_fields = [
            "intro",
        ]

@api.get("/pages/{page_id}/", response=BlogPageSchema)
def get_page(request: "HttpRequest", page_id: int):
    return get_object_or_404(Page, id=page_id).specific
```

This works well, with the endpoint now returning generic `Page` fields and the `BlogPage` introduction.
But for sites where all page content is served via an API, it could become tedious to create new endpoints for every page type.

### Combining multiple schemas

> To reflect our response may return multiple page types, we use the `Union` type hint to combine multiple schemas.

Reviewer 2: ```suggestion
To reflect that our response may return multiple page types, we use the `Union` type hint to combine multiple schemas.
```
